From 4d4d7f6319b0a0b63c5c23728dea50be8eca0dc7 Mon Sep 17 00:00:00 2001
From: Daniel Johnson <danieljohnson2@fastmail.com>
Date: Thu, 18 May 2023 17:00:13 -0400
Subject: [PATCH] Handle obsolete runners more safely.

Resolves #4863

In case one winds up in the sidebar, it will have no actions.
---
 lutris/gui/widgets/sidebar.py | 7 ++++++-
 lutris/runners/__init__.py    | 5 ++++-
 2 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/lutris/gui/widgets/sidebar.py b/lutris/gui/widgets/sidebar.py
index b54260ebcb..00e962207c 100644
--- a/lutris/gui/widgets/sidebar.py
+++ b/lutris/gui/widgets/sidebar.py
@@ -15,6 +15,7 @@
 from lutris.gui.dialogs.runner_install import RunnerInstallDialog
 from lutris.gui.widgets.utils import has_stock_icon
 from lutris.installer.interpreter import ScriptInterpreter
+from lutris.runners import InvalidRunner
 from lutris.services import SERVICES
 from lutris.services.base import AuthTokenExpired, BaseService
 
@@ -212,7 +213,11 @@ def get_actions(self):
 
         # Creation is delayed because only installed runners can be imported
         # and all visible boxes should be installed.
-        self.runner = runners.import_runner(self.id)()
+        try:
+            self.runner = runners.import_runner(self.id)()
+        except InvalidRunner:
+            return entries
+
         if self.runner.multiple_versions:
             entries.append((
                 "system-software-install-symbolic",
diff --git a/lutris/runners/__init__.py b/lutris/runners/__init__.py
index 0f5cb1e7f7..fb59191b26 100644
--- a/lutris/runners/__init__.py
+++ b/lutris/runners/__init__.py
@@ -120,7 +120,10 @@ def get_runner_human_name(runner_name):
     names."""
     if runner_name:
         if runner_name not in _cached_runner_human_names:
-            _cached_runner_human_names[runner_name] = import_runner(runner_name)().human_name
+            try:
+                _cached_runner_human_names[runner_name] = import_runner(runner_name)().human_name
+            except InvalidRunner:
+                _cached_runner_human_names[runner_name] = runner_name  # an obsolete runner
         return _cached_runner_human_names[runner_name]
 
     return ""
